cupy/cuda/pinned_memory.pyx

```diff
@@ -0,0 +1,270 @@
+
+        self._shape[0] = size
+        self._strides[0] = 1
+
+        buffer.buf = <void*>self.ptr
+        buffer.format = 'b'
+        buffer.internal = NULL
+        buffer.itemsize = 1
+        buffer.len = size
+        buffer.ndim = 1
+        buffer.obj = self
+        buffer.readonly = 0
+        buffer.shape = self._shape
+        buffer.strides = self._strides
+        buffer.suboffsets = NULL
+
+    def __releasebuffer__(self, Py_buffer *buffer):
+        pass
+
+    def __getsegcount__(self, Py_ssize_t *lenp):
+        if lenp != NULL:
+            lenp[0] = self.size()
+        return 1
+
+    def __getreadbuffer__(self, Py_ssize_t idx, void **p):
+        if idx != 0:
+            raise SystemError("accessing non-existent buffer segment")
+        p[0] = <void*>self.ptr
+        return self.size()
+
+    def __getwritebuffer__(self, Py_ssize_t idx, void **p):
+        if idx != 0:
+            raise SystemError("accessing non-existent buffer segment")
+        p[0] = <void*>self.ptr
+        return self.size()
+
+
+cpdef PinnedMemoryPointer _malloc(Py_ssize_t size):
+    mem = PinnedMemory(size, runtime.hostAllocPortable)
+    return PinnedMemoryPointer(mem, 0)
+
+
+cdef object _current_allocator = _malloc
+
+
+cpdef PinnedMemoryPointer alloc_pinned_memory(Py_ssize_t size):
+    """Calls the current allocator.
+
+    Use :func:`~cupy.cuda.set_pinned_memory_allocator` to change the current
+    allocator.
+
+    Args:
+        size (int): Size of the memory allocation.
+
+    Returns:
+        ~cupy.cuda.PinnedMemoryPointer: Pointer to the allocated buffer.
+
+    """
+    return _current_allocator(size)
+
+
+cpdef set_pinned_memory_allocator(allocator=_malloc):
+    """Sets the current allocator.
+
+    Args:
+        allocator (function): CuPy pinned memory allocator. It must have the
+            same interface as the :func:`cupy.cuda.alloc_alloc_pinned_memory`
+            function, which takes the buffer size as an argument and returns
+            the device buffer of that size.
+
+    """
+    global _current_allocator
+    _current_allocator = allocator
+
+
+cdef class PooledPinnedMemory(PinnedMemory):
+
+    """Memory allocation for a memory pool.
+
+    The instance of this class is created by memory pool allocator, so user
+    should not instantiate it by hand.
+
+    """
+
+    def __init__(self, PinnedMemory mem, pool):
+        self.ptr = mem.ptr
+        self.size = mem.size
+        self.pool = pool
+
+    def __dealloc__(self):
+        if self.ptr != 0:
+            self.free()
+
+    cpdef free(self):
+        """Frees the memory buffer and returns it to the memory pool.
+
+        This function actually does not free the buffer. It just returns the
+        buffer to the memory pool for reuse.
+
+        """
+        pool = self.pool()
+        if pool and self.ptr != 0:
+            pool.free(self.ptr, self.size)
+        self.ptr = 0
+        self.size = 0
+
+
+cdef class PinnedMemoryPool:
```

Comment:
The memory pool of CDUA memory has same problem.
When we find better memory pooling strategy, we should change behavior.